File: Omega/src/com/saggitt/omega/preferences/GridSizePreference.java
/*
 *  This file is part of Omega Launcher
 *  Copyright (c) 2021   Saul Henriquez
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.saggitt.omega.preferences;

import android.content.Context;
import android.util.AttributeSet;

import androidx.core.util.Pair;
import androidx.preference.DialogPreference;

import com.android.launcher3.R;
import com.android.launcher3.Utilities;

public class GridSizePreference extends DialogPreference {
    private GridSize2D gridSize;
    private Pair defaultSize;

    public GridSizePreference(Context context) {
        this(context, null, 0);
    }

    public GridSizePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        updateSummary();
    }

    public GridSizePreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        updateSummary();
    }

    public Pair getSize() {

        gridSize = Utilities.getOmegaPrefs(getContext()).getGridSize();
        defaultSize = new Pair(gridSize.getNumRowsOriginal(), gridSize.getNumColumnsOriginal());

        int rows = gridSize.fromPref(gridSize.getNumRows(), (int) defaultSize.first);
        int columns = gridSize.fromPref(gridSize.getNumColumns(), (int) defaultSize.second);
        return new Pair(rows, columns);
    }


    public void setSize(int rows, int columns) {
        gridSize.setNumRowsPref(gridSize.toPref(rows, (int) defaultSize.first));
        gridSize.setNumColumnsPref(gridSize.toPref(columns, (int) defaultSize.second));
        updateSummary();
    }


    private void updateSummary() {
        Pair value = getSize();
        setSummary(String.format("%d x %d", value.first, value.second));
    }

    @Override
    public int getDialogLayoutResource() {
        return R.layout.pref_dialog_grid_size;
    }
}

File: Omega/src/com/saggitt/omega/preferences/custom/GridSizePreference.kt
package com.saggitt.omega.preferences.custom

import android.content.Context
import android.util.AttributeSet
import androidx.annotation.StringRes
import androidx.preference.DialogPreference
import com.android.launcher3.R
import com.android.launcher3.Utilities
import com.saggitt.omega.preferences.BasePreferences
import com.saggitt.omega.util.JavaField

class GridSizePreference(context: Context, attrs: AttributeSet?) :
    DialogPreference(context, attrs) {
    val gridSize = Utilities.getOmegaPrefs(context).desktopGridSize
    val defaultSize by lazy { Pair(gridSize.numColumnsOriginal, gridSize.numRowsOriginal) }

    init {
        updateSummary()
    }

    fun getSize(): Pair<Int, Int> {
        val rows = gridSize.fromPref(gridSize.numColumns, defaultSize.first)
        val columns = gridSize.fromPref(gridSize.numRows, defaultSize.second)
        return Pair(rows, columns)
    }

    fun setSize(rows: Int, columns: Int) {
        gridSize.numColumnsPref.onSetValue(gridSize.toPref(rows, defaultSize.first))
        gridSize.numRowsPref.onSetValue(gridSize.toPref(columns, defaultSize.second))
        updateSummary()
    }

    private fun updateSummary() {
        val value = getSize()
        summary = "${value.first}x${value.second}"
    }

    override fun getDialogLayoutResource() = R.layout.pref_dialog_grid_size
}

open class GridSize(
    @StringRes val titleId: Int,
    val numColumnsPref: BasePreferences.IdpIntPref,
    columnsKey: String,
    targetObject: Any,
    private val onChangeListener: () -> Unit
) {
    var numColumns by JavaField<Int>(targetObject, columnsKey)
    val numColumnsOriginal by JavaField<Int>(targetObject, "${columnsKey}Original")
    protected val onChange = {
        applyCustomization()
        onChangeListener.invoke()
    }

    init {
        applyNumColumns()
    }

    protected open fun applyCustomization() {
        applyNumColumns()
    }

    private fun applyNumColumns() {
        numColumns = fromPref(numColumnsPref.onGetValue(), numColumnsOriginal)
    }

    fun fromPref(value: Int, default: Int) = if (value != 0) value else default
    fun toPref(value: Int, default: Int) = if (value != default) value else 0
}

class GridSize2D(
    @StringRes titleId: Int,
    numColumnsPref: BasePreferences.IdpIntPref,
    val numRowsPref: BasePreferences.IdpIntPref,
    columnsKey: String,
    rowsKey: String,
    targetObject: Any,
    onChangeListener: () -> Unit
) : GridSize(
    titleId,
    numColumnsPref,
    columnsKey,
    targetObject,
    onChangeListener
) {
    var numRows by JavaField<Int>(targetObject, rowsKey)
    val numRowsOriginal by JavaField<Int>(targetObject, "${rowsKey}Original")

    init {
        applyNumRows()
    }

    override fun applyCustomization() {
        super.applyCustomization()
        applyNumRows()
    }

    private fun applyNumRows() {
        numRows = fromPref(numRowsPref.onGetValue(), numRowsOriginal)
    }
}